File: atcoder/abs/04-coins/go/main.go
// Copyright (c) 2023 Furukawa, Atsushi <atsushifx@aglabo.com>
//
// This software is released under the MIT License.
// https://opensource.org/licenses/MIT

package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

// Global Variables
const IOBUFFSIZE = 10 ^ 6

var Stdin = bufio.NewScanner(os.Stdin)
var Stdout = bufio.NewWriter(os.Stdout)

// atcoder lib
func getNext() string {
	Stdin.Scan()
	return Stdin.Text()
}

func getNextInt() int {
	var i int
	var e error
	i, e = strconv.Atoi(getNext())
	if e != nil {
		panic(e)
	}
	return i
}

func getNums(size int) []int {
	var buff []int = make([]int, size)
	for i := range buff {
		buff[i] = getNextInt()
	}
	return buff
}

// initialize
func initialize() {
	stdinBuffer := make([]byte, IOBUFFSIZE)
	Stdin.Buffer(stdinBuffer, IOBUFFSIZE)
	Stdin.Split(bufio.ScanWords)
}

// main
func main() {
	initialize()
	defer Stdout.Flush()
	solve()
	Stdout.Flush()
}

// problem solver
func solve() {
	// input
	var a int = getNextInt()
	var b int = getNextInt()
	var c int = getNextInt()

	var x = getNextInt()

	// solve
	var count int = 0

	for i := 0; i <= a; i++ {
		for j := 0; j <= b; j++ {
			for k := 0; k <= c; k++ {
				var sum = 500*i + 100*j + 50*k
				if sum == x {
					count++
					// fmt.Fprintf(Stdout, "[%d %d %d]\n", i, j, k)
				}
			}
		}
	}

	// Out
	fmt.Fprintln(Stdout, count)

}

File: atcoder/abs/04-coins/typescript/main.ts
// Copyright (c) 2023 Furukawa, Atsushi <atsushifx@aglabo.com>
//
// This software is released under the MIT License.
// https://opensource.org/licenses/MIT

/*
 * atcoder abs : coins
 */
import * as fs from "fs";

// stdin
const inputs: string = (fs.readFileSync(process.stdin.fd, "utf8"))
const inputArray: string[] = inputs.split(/\s/)
let inputIndex = 0;

// atcoder lib:stdin
function getNext(): string {
	return inputArray[inputIndex++];
}

function getNextNum(): number {
	return Number(getNext());
}

function getNums(N: number): number[] {
	const arr: number[] = [];
	for (let i: number = 0; i < N; ++i) {
		arr[i] = getNextNum();
	}
	return arr;
}

// lib:stdout
let outputBuffer: string = ""
function flush(): void {
	console.log(outputBuffer);
}

// exec main
solve();


// solve: solve problem main
function solve() {
	// Input
	const [a, b, c] = getNums(3);
	const x = getNextNum()

	// Answer
	let count: number = 0
	for (let i: number = 0; i <= a; ++i) {
		for (let j: number = 0; j <= b; ++j) {
			for (let k: number = 0; k <= c; ++k) {
				let sum: number = 500 * i + 100 * j + 50 * k
				if (sum == x) {
					count += 1
				}
				if (sum > x) {
					break
				}
			}
		}
	}

	// Out
	console.log(count)
}
